src/libaktualizr/storage/sqlstorage.cc

```diff
@@ -1028,67 +1024,97 @@ class SQLTargetWHandle : public StorageTargetWHandle {
+    std::lock_guard<std::mutex> lock(sql_mutex);
+
+    if (db_.get_rc() != SQLITE_OK) {
+      LOG_ERROR << "Can't open database: " << db_.errmsg();
+      throw StorageTargetWHandle::WriteError("could not save file " + target_.filename() + " to sql storage");
+    }
+
+    written_size_ = start_from;
+  }
   SQLite3Guard db_;
-  const std::string filename_;
-  size_t expected_size_;
+  Uptane::Target target_;
   bool closed_;
   sqlite3_blob* blob_;
   sqlite3_int64 row_id_;
 };
 
-std::unique_ptr<StorageTargetWHandle> SQLStorage::allocateTargetFile(bool from_director, const std::string& filename,
-                                                                     size_t size) {
+std::unique_ptr<StorageTargetWHandle> SQLStorage::allocateTargetFile(bool from_director, const Uptane::Target& target) {
   (void)from_director;
-
-  return std::unique_ptr<StorageTargetWHandle>(new SQLTargetWHandle(*this, filename, size));
+  return std::unique_ptr<StorageTargetWHandle>(new SQLTargetWHandle(*this, target));
 }
 
 class SQLTargetRHandle : public StorageTargetRHandle {
  public:
-  SQLTargetRHandle(const SQLStorage& storage, const std::string& filename)
-      : db_(storage.dbPath()), filename_(filename), size_(0), read_size_(0), closed_(false), blob_(nullptr) {
+  SQLTargetRHandle(const SQLStorage& storage, const Uptane::Target& target)
+      : db_path_(storage.dbPath()),
+        db_(db_path_),
+        target_(target),
+        size_(0),
+        read_size_(0),
+        closed_(false),
+        blob_(nullptr) {
     StorageTargetRHandle::ReadError exc("could not read file " + filename_ + " from sql storage");
 
     if (db_.get_rc() != SQLITE_OK) {
       LOG_ERROR << "Can't open database: " << db_.errmsg();
       throw exc;
     }
 
-    if (!db_.beginTransaction()) {
-      throw exc;
-    }
-    auto statement =
-        db_.prepareStatement<std::string>("SELECT rowid, real_size FROM target_images WHERE filename = ?;", filename);
-
-    int err = statement.step();
-    if (err == SQLITE_DONE) {
-      LOG_ERROR << "No such file in db: " + filename_;
-      throw exc;
-    }
-    if (err != SQLITE_ROW) {
-      LOG_ERROR << "Statement step failure: " << db_.errmsg();
-      throw exc;
-    }
+    auto statement = db_.prepareStatement<std::string>(
+        "SELECT rowid, real_size, sha256, sha512 FROM target_images WHERE filename = ?;", target.filename());
 
-    auto row_id = statement.get_result_col_int(0);
-    if (sqlite3_blob_open(db_.get(), "main", "target_images", "image_data", row_id, 0, &blob_) != SQLITE_OK) {
-      LOG_ERROR << "Could not open blob: " << db_.errmsg();
-      throw exc;
-    }
-    size_ = static_cast<size_t>(sqlite3_blob_bytes(blob_));
+    int statement_state;
+    while ((statement_state = statement.step()) == SQLITE_ROW) {
+      auto sha256 = statement.get_result_col_str(2);
+      auto sha512 = statement.get_result_col_str(3);
+      if (!sha256 && !sha512) {
```

Comment:
`!sha256 || !sha512`?

Reply:
No, we skip if target don't have both hashes

Reply:
Oh ok. But the code that follows is calling `*` on both. So if one is defined and not the other, it will pass the check and then crash when dereferencing the other.

Reply:
No, because it is NOT NULL column. So it  can be empty string  or hash. But in this case `if (!sha256 && !sha512) {` is incorrect, we should check for empty string. So it should be `if ((*sha256).empty() && (*sha512).empty()) {`